force-app/main/default/classes/DMLOperations.cls

```diff
@@ -178,41 +205,103 @@ public with sharing class DMLOperations {
      * Example Contact.LastName = 'Jane' Link to Account.Name = 'Jane'
 	 * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
 	 */
-	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
-		// Loop through each Contact
-
-		// Extract the Account Name from Contact's LastName
 
-		// Get the account based on the name or create one if it doesn't exist
-
-		// Link the Contact to its Account using the Account Id
-
-		// Upsert the Contacts
-	}
+        public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
+            Map<String, Account> accountMap = new Map<String, Account>();
+            List<Account> accountsToInsert = new List<Account>();
+            Set<String> accountNames = new Set<String>();
+    
+            for (Contact contact : contactsToUpsert) {
+                String accountName = contact.LastName;
+                if (String.isNotBlank(accountName)) {
+                    accountNames.add(accountName);
+                }
+            }
+    
+            List<Account> existingAccounts = [SELECT Id, Name FROM Account WHERE Name IN :accountNames WITH USER_MODE];
+            for (Account acc : existingAccounts) {
+                accountMap.put(acc.Name, acc);
+            }
+    
+            for (Contact contact : contactsToUpsert) {
+                String accountName = contact.LastName;
+                if (String.isNotBlank(accountName) && !accountMap.containsKey(accountName)) {
+                    Account newAcc = new Account(Name = accountName);
+                    accountsToInsert.add(newAcc);
+                    accountMap.put(accountName, newAcc);
+                }
+            }
+    
+            if (!accountsToInsert.isEmpty()) {
+                insert as user accountsToInsert;
+            }
+    
+            for (Contact contact : contactsToUpsert) {
+                String accountName = contact.LastName;
+                if (String.isNotBlank(accountName) && accountMap.containsKey(accountName)) {
+                    contact.AccountId = accountMap.get(accountName).Id;
+                }
+            }
+            upsert as user contactsToUpsert;
```

Comment:
Really amazing job! This is not an easy problem and you executed it beautifully! Very well structured, clean code.